api/topic.go

```diff
@@ -65,6 +62,127 @@ func (api *API) getTopicPrivateHandler(w http.ResponseWriter, req *http.Request)
 }
 
+// getSubtopicsPublicHandler is a handler that gets a topic by its id from MongoDB
+func (api *API) getSubtopicsPublicHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	vars := mux.Vars(req)
+	id := vars["id"]
+	logdata := log.Data{
+		"request_id": ctx.Value(dprequest.RequestIdKey),
+		"topic_id":   id,
+		"function":   "getSubtopicsPublicHandler",
+	}
+
+	// get topic from mongoDB by id
+	topic, err := api.dataStore.Backend.GetTopic(id)
+	if err != nil {
+		// no topic found to retrieve the subtopics from
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// User is not authenticated and hence has only access to current sub document(s)
+	var result models.PublicSubtopics
+
+	if topic.Current == nil {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if len(topic.Current.SubtopicIds) == 0 {
+		// no subtopics exist for the requested ID
+		handleError(ctx, w, apierrors.ErrNotFound, logdata)
+		return
+	}
+
+	for _, subTopicID := range topic.Current.SubtopicIds {
+		// get sub topic from mongoDB by subTopicID
+		topic, err := api.dataStore.Backend.GetTopic(subTopicID)
+		if err != nil {
+			logdata["missing subtopic for id"] = subTopicID
+			log.Event(ctx, err.Error(), log.ERROR, logdata)
+			continue
+		}
+		result.PublicItems = append(result.PublicItems, topic.Current)
+		result.TotalCount++
+	}
+	if result.TotalCount == 0 {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if err := WriteJSONBody(ctx, result, w, logdata); err != nil {
+		return
+	}
+	log.Event(ctx, "request successful", log.INFO, logdata) // NOTE: name of function is in logdata
+}
+
+// getSubtopicsPrivateHandler is a handler that gets a topic by its id from MongoDB
+func (api *API) getSubtopicsPrivateHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	vars := mux.Vars(req)
+	id := vars["id"]
+	logdata := log.Data{
+		"request_id": ctx.Value(dprequest.RequestIdKey),
+		"topic_id":   id,
+		"function":   "getSubtopicsPrivateHandler",
+	}
+
+	// get topic from mongoDB by id
+	topic, err := api.dataStore.Backend.GetTopic(id)
+	if err != nil {
+		// no topic found to retrieve the subtopics from
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// User has valid authentication to get raw full topic document(s)
+	var result models.PrivateSubtopics
+
+	if topic.Current == nil {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if topic.Next == nil {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if len(topic.Current.SubtopicIds) == 0 {
+		// no subtopics exist for the requested ID
+		handleError(ctx, w, apierrors.ErrNotFound, logdata)
+		return
+	}
+
+	if len(topic.Current.SubtopicIds) != len(topic.Next.SubtopicIds) {
+		// no subtopics exist for the requested ID
+		handleError(ctx, w, apierrors.ErrNotFound, logdata)
+		return
+	}
+
+	for _, subTopicID := range topic.Current.SubtopicIds {
```

Comment:
I am not sure how the current and next docs are to work in the topic API. Assuming it's the same as the usage in the dataset API, the private endpoint would return data from the next document as it's what's due to be published.

Reply:
this does return next and current together.
see : models.PrivateSubtopics
which contains a []*TopicResponse
which contains Next & Current.